### PR TITLE
Update Drain Cleaner to work with Node Pools

### PR DESCRIPTION
**Pull request by ShazaAldawamneh**
new method to handle pods based on labels rather than pod name for Drain Cleaner and creating and editing tests

### CODE REVIEW

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 1]
```diff
@@ -23,15 +23,13 @@
...
 
 @Path("/drainer")
 public class ValidatingWebhook {
     private static final Logger LOG = LoggerFactory.getLogger(ValidatingWebhook.class);
 
-    private static final Pattern ZOOKEEPER_PATTERN = Pattern.compile(".+(-zookeeper-)[0-9]+");
-    private static final Pattern KAFKA_PATTERN = Pattern.compile(".+(-kafka-)[0-9]+");
-
+    private static final String KIND_LABEL_VALUE = "Kafka";
+    private static final String KIND_LABEL_KEY = "strimzi.io/kind";
```
Comment (scholzj) [inline src/main/java/io/strimzi/ValidatingWebhook.java:32]: I don't think the `strimzi.io/kind` label works. That would capture also all the Deployment managed pods. You need to target the `strimzi.io/name` label I think and match it against something like `.+(-zookeeper)` and `.+(-kafka)`.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 2]
```diff
@@ -65,6 +63,10 @@ private ObjectMeta extractEvictionMetadata(AdmissionRequest request)    {
         }
     }
 
+    private boolean matchingLabel(Map<String, String> labels) {
+        return labels.containsKey(KIND_LABEL_KEY) && labels.get(KIND_LABEL_KEY).equals(KIND_LABEL_VALUE);
```
Comment (scholzj) [inline src/main/java/io/strimzi/ValidatingWebhook.java:67]: Can `labels` be null`? If not, you can simply do this:
```
return KIND_LABEL_VALUE.equals(labels.get(KIND_LABEL_KEY))
```

If it can be null, you should do something like:
```
return labels != null && KIND_LABEL_VALUE.equals(labels.get(KIND_LABEL_KEY))
```

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 3]
```diff
@@ -54,7 +54,7 @@ public void setup() {
 
     @Test
     public void testEviction() {
-        String podName = "my-cluster-kafka-1";
+        String podName = "my-cluster";
```
Comment (scholzj) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:57]: Why did you changed the name here? The original name was one of the expected ones. The new one is not.

File: src/main/java/io/strimzi/ValidatingWebhook.java (hunk 1 above)
Comment (fvaleri) [reply src/main/java/io/strimzi/ValidatingWebhook.java:32]: > I don't think the strimzi.io/kind label works. 

Correct, that label also includes the entity operator pod for example.

Instead, the value of `strimzi.io/name` is created with the pattern $CLUSTER_NAME-kafka or $CLUSTER_NAME-zookeeper. While $CLUSTER_NAME part is variable, the second part (-kafka and -zookeeper) is constant and that's what you can use for matching.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 4]
```diff
@@ -74,12 +78,10 @@ public AdmissionReview webhook(AdmissionReview review) {
         AdmissionRequest request = review.getRequest();
         ObjectMeta evictionMetadata = extractEvictionMetadata(request);
 
-        if (evictionMetadata != null)   {
-            if (drainZooKeeper && ZOOKEEPER_PATTERN.matcher(evictionMetadata.getName()).matches()
-                    || drainKafka && KAFKA_PATTERN.matcher(evictionMetadata.getName()).matches()) {
+        if (evictionMetadata != null) {
+            if ((drainKafka || drainZooKeeper) && (matchingLabel(evictionMetadata.getLabels()))) {
```
Comment (scholzj) [inline src/main/java/io/strimzi/ValidatingWebhook.java:82]: I think this does not distinguish between Kafka and ZooKeeper being enabled. For example, when you enable it only for Kafka (i.e. `drainZooKeeper` is `false`), it would still pass the first part of the condition (`drainKafka || drainZooKeeper`) and handle the ZooKeeper pod. So this seems wrong.

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 5]
```diff
@@ -147,13 +148,14 @@ public void testEvictionOnPodWithExistingAnnotations() {
 
     @Test
     public void testWrongName() {
```
Comment (scholzj) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:150]: Maybe you can rename the test as it now tests wrong label rather then wrong name?

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 6]
```diff
@@ -208,6 +210,7 @@ public void testNoKindLabel() {
         assertThat(reviewResponse.getResponse().getAllowed(), is(true));
         verify(podResource, times(1)).get();
         verify(podResource, never()).patch((Pod) any());
+
```
Comment (scholzj) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:213]: Not needed?

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 7]
```diff
@@ -178,15 +180,15 @@ public void testKafkaAndZooKeeperFilters() {
...
+        AdmissionReview reviewResponse = webhook.webhook(reviewRequest("my-cluster", true));
 
         assertThat(reviewResponse.getResponse().getUid(), is("SOME-UUID"));
         assertThat(reviewResponse.getResponse().getAllowed(), is(true));
         verify(podResource, never()).get();
         verify(podResource, never()).patch((Pod) any());
 
         // Test it for ZooKeeper
-        reviewResponse = webhook.webhook(reviewRequest("my-cluster-zookeeper-1", true));
+        reviewResponse = webhook.webhook(reviewRequest("my-cluster", true));
```
Comment (scholzj) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:191]: How does this test the different patters? You do not seem to change the labels.

File: src/test/java/io/strimzi/systemtest/DrainCleanerST.java [hunk 8]
```diff
@@ -47,19 +57,26 @@ void testEvictionRequestOnKafkaPod() {
     @Test
     void testEvictionRequestOnRandomPod() {
```
Comment (fvaleri) [inline src/test/java/io/strimzi/systemtest/DrainCleanerST.java:58]: Why we need to change this test?

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 9]
```diff
@@ -78,6 +78,7 @@ public void testV1Beta1Eviction() {
                 .withNewMetadata()
                     .withName(podName)
                     .withNamespace("my-namespace")
+                    .withLabels(Collections.singletonMap("strimzi.io/name", "my-cluster-kafka"))
```
Comment (scholzj) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:81]: I would expect to see tests without any labels, with some completely different label, with correct label with wrong value and finally with correct label and correct Kafka and ZooKeeper value.

In addition, a test with Kafka label but disabled Kafka draining and ZooKeeper label and disabled ZooKeeper draining would be good.

**Review by scholzj — COMMENTED** [review]
I tjhink Fede's comments are valid. I added a bit more as to what you can test in the unit tests.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 10]
```diff
@@ -99,13 +103,12 @@ public AdmissionReview webhook(AdmissionReview review) {
         } else {
             LOG.warn("Weird, this does not seem to be an Eviction webhook.");
         }
-
         return new AdmissionReviewBuilder()
-                .withNewResponse()
```
[the remark this answers is not in the document]
Comment (ShazaAldawamneh) [reply src/main/java/io/strimzi/ValidatingWebhook.java:104]: I didn't added/removed that line, can you clarify more ?

[the remark this answers is not in the document]
Comment (ShazaAldawamneh) [reply src/main/java/io/strimzi/ValidatingWebhook.java:104]: @fvaleri

File: src/test/java/io/strimzi/systemtest/DrainCleanerST.java (hunk 8 above)
Comment (ShazaAldawamneh) [reply src/test/java/io/strimzi/systemtest/DrainCleanerST.java:58]: because it need to match with the new code. 
as it was failing

File: src/main/java/io/strimzi/ValidatingWebhook.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (fvaleri) [reply src/main/java/io/strimzi/ValidatingWebhook.java:104]: It's not like you removed anything, but the code alignment was changed (maybe it was the IDE that did this automatically). The diff presents this as code removed and added back with additional 4 spaces on each line. Just revert this change by removing these extra spaces in lines 107-111 and you'll be fine :)

File: src/test/java/io/strimzi/systemtest/DrainCleanerST.java (hunk 8 above)
Comment (fvaleri) [reply src/test/java/io/strimzi/systemtest/DrainCleanerST.java:58]: Here you can simply create an overload version of `createStatefulSetAndPDBWithWait` which takes just the `stsName` as input parameter and call the new version passing `stsName` an empty map of labels. That way, you can reuse the old `testEvictionRequestOnRandomPod` code.

Additionally, using plural is more appropriate for a map, so I would use `labels` instead of `label` everywhere.

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 11]
```diff
@@ -146,14 +147,117 @@ public void testEvictionOnPodWithExistingAnnotations() {
     }
 
     @Test
-    public void testWrongName() {
-        String podName = "my-cluster-connect-1";
+    public void testWrongLabel() {
+        String podName = "my-cluster-kafka-1";
+        Map label = Collections.singletonMap("mylabel", "foo");
         when(podResource.get()).thenReturn(mockedPod(podName, true, false));
```
Comment (fvaleri) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:153]: Pod name is now irrelevant and can be hard coded. Instead, the set of labels is what you want to test against the various webhook configurations.

Whatever label comes in a `reviewRequest`, should also be in the `mockedPod`, so it could help to factor out the map of labels and pass it as parameter to both methods.

File: src/test/java/io/strimzi/systemtest/DrainCleanerST.java [hunk 12]
```diff
@@ -27,39 +28,60 @@ public class DrainCleanerST extends AbstractST {
...
+        LOGGER.info("Creating eviction request to the pod.");
+        ObjectMeta meta = kubeClient().listPodsByPrefixInName(Constants.NAMESPACE, stsName).get(0).getMetadata();
+
+        kubeClient().getClient().pods().inNamespace(Constants.NAMESPACE).withName(meta.getName())
+            .evict(new EvictionBuilder()
+                .withMetadata(meta)
+                .withDeleteOptions(new DeleteOptionsBuilder()
+                    .withGracePeriodSeconds(0L)
+                    .build())
+                .build());
```
Comment (scholzj) [inline src/test/java/io/strimzi/systemtest/DrainCleanerST.java:50]: Why was this changed? The original code should IMHO keep working with simple `evict()`.

File: src/test/java/io/strimzi/systemtest/DrainCleanerST.java [hunk 13]
```diff
@@ -27,39 +28,60 @@ public class DrainCleanerST extends AbstractST {
...
-        createStatefulSetAndPDBWithWait(stsName);
-
-        LOGGER.info("Creating eviction request to the pod");
-        String podName = kubeClient().listPodsByPrefixInName(Constants.NAMESPACE, stsName).get(0).getMetadata().getName();
-        kubeClient().getClient().pods().inNamespace(Constants.NAMESPACE).withName(podName).evict();
-
+        final String stsName = "my-cluster-bar";
+        final Map<String, String> labels = Map.of(
+                "app", stsName,
+                "strimzi.io/kind", "Kafka"
```
Comment (scholzj) [inline src/test/java/io/strimzi/systemtest/DrainCleanerST.java:65]: I would remove this label from an random pod.

File: src/test/java/io/strimzi/systemtest/DrainCleanerST.java [hunk 14]
```diff
@@ -27,39 +28,60 @@ public class DrainCleanerST extends AbstractST {
...
+        LOGGER.info("Creating eviction request to the pod.");
+        ObjectMeta meta = kubeClient().listPodsByPrefixInName(Constants.NAMESPACE, stsName).get(0).getMetadata();
+
+        kubeClient().getClient().pods().inNamespace(Constants.NAMESPACE).withName(meta.getName())
+            .evict(new EvictionBuilder()
+                .withMetadata(meta)
+                .withDeleteOptions(new DeleteOptionsBuilder()
+                    .withGracePeriodSeconds(0L)
+                    .build())
+                .build());
```
Comment (scholzj) [inline src/test/java/io/strimzi/systemtest/DrainCleanerST.java:79]: Same as above ... why is the original `evict()` not sufficient?

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 15]
```diff
@@ -54,13 +54,17 @@ public void setup() {
 
     @Test
     public void testEviction() {
-        String podName = "my-cluster-kafka-1";
-        when(podResource.get()).thenReturn(mockedPod(podName, true, false));
+        String podName = "my-cluster-foo";
```
Comment (fvaleri) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:57]: You can get rid of the `podName` parameter in `mockedPod` and `reviewRequest`. Simply set a constant like `my-cluster-kafka-1`.

You can also get rid of the `kindLabel` boolean parameter in `mockedPod`, because it's now part of the input map of labels.

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 16]
```diff
@@ -148,14 +163,18 @@ public void testEvictionOnPodWithExistingAnnotations() {
...
+                "strimzi.io/name", "my-cluster-foo"
+        );
+
+        when(podResource.get()).thenReturn(mockedPod(podName, true, false, labels));
         ArgumentCaptor<Pod> podCaptor = ArgumentCaptor.forClass(Pod.class);
         when(podResource.patch(podCaptor.capture())).thenReturn(new Pod());
 
         ValidatingWebhook webhook = new ValidatingWebhook(client, true, true);
-        AdmissionReview reviewResponse = webhook.webhook(reviewRequest(podName, false, label));
+        AdmissionReview reviewResponse = webhook.webhook(reviewRequest(podName, false, labels));
```
Comment (fvaleri) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:177]: We are missing assertions about annotations in a bunch of tests, including this one.

**Review by fvaleri — COMMENTED** [review]
I think we are almost there. Leaved few additional comments.

File: src/test/java/io/strimzi/systemtest/DrainCleanerST.java (hunk 12 above)
Comment (fvaleri) [reply src/test/java/io/strimzi/systemtest/DrainCleanerST.java:50]: @scholzj looks like it doesn't work, at least on my env.

Looks like the ObjectMeta of the pod to be evicted is required, otherwise you don't get a match and this is what the webhook prints out (the 2nd should be a match):

```sh
2023-06-20 07:30:03,867 INFO  [io.str.ValidatingWebhook] (executor-thread-1) Received eviction event which does not match any relevant pods.
2023-06-20 07:31:16,672 INFO  [io.str.ValidatingWebhook] (executor-thread-0) Received eviction event which does not match any relevant pods.
```

Comment (fvaleri) [reply src/test/java/io/strimzi/systemtest/DrainCleanerST.java:50]: I'll do that.

Comment (scholzj) [reply src/test/java/io/strimzi/systemtest/DrainCleanerST.java:50]: I get this in the real drain:

```
2023-06-20 09:45:32,976 INFO [io.str.ValidatingWebhook] (executor-thread-4) Received eviction event which does not match any relevant pods.
```

So I think the test needs to do the same it did before. And the code needs to get the pod first based on the name and then check the labels on the pod.

Comment (fvaleri) [reply src/test/java/io/strimzi/systemtest/DrainCleanerST.java:50]: Yeah, I did a test on my end and I can confirm.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 17]
```diff
@@ -28,10 +28,9 @@
 @Path("/drainer")
 public class ValidatingWebhook {
     private static final Logger LOG = LoggerFactory.getLogger(ValidatingWebhook.class);
-
-    private static final Pattern ZOOKEEPER_PATTERN = Pattern.compile(".+(-zookeeper-)[0-9]+");
-    private static final Pattern KAFKA_PATTERN = Pattern.compile(".+(-kafka-)[0-9]+");
-
+    private static final Pattern ZOOKEEPER_PATTERN = Pattern.compile(".+-zookeeper");
+    private static final Pattern KAFKA_PATTERN = Pattern.compile(".+-kafka");
+    private static final String STRIMZI_LABEL_KEY = "strimzi.io/name";
```
[the remark this answers is not in the document]
Comment (fvaleri) [reply src/main/java/io/strimzi/ValidatingWebhook.java:34]: This hasn't been addressed.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 18]
```diff
@@ -74,32 +83,35 @@ public AdmissionReview webhook(AdmissionReview review) {
...
-                    namespace = request.getNamespace();
-                }
-
-                if (name == null || namespace == null) {
-                    LOG.warn("Failed to decode pod name or namespace from the eviction webhook (pod: {}, namespace: {})", name, namespace);
+        if (evictionMetadata != null && evictionMetadata.getNamespace() != null) {
+            String name = evictionMetadata.getName();
+            String namespace = evictionMetadata.getNamespace();
+
+            Pod pod = client.pods().inNamespace(namespace).withName(name).get();
```
Comment (fvaleri) [inline src/main/java/io/strimzi/ValidatingWebhook.java:90]: You should search for the pod only after checking that both name and namespace are not null or blank. Below we also get the namespace from the request in case it is missing from the eviction, but this code is never executed.

After these checks, you can get the pod, match it against our rules and pass it as parameter to `annotatePodForRestart` to avoid doing another API call.

**Review by fvaleri — CHANGES_REQUESTED** [review]
Hi @ShazaAldawamneh, I think there are still some issues to address. Let me know if something is not clear.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 19]
```diff
@@ -74,32 +83,35 @@ public AdmissionReview webhook(AdmissionReview review) {
...
+            if (pod != null) {
+                if (matchingLabel(pod.getMetadata().getLabels())) {
+                    if (namespace == null) {
+                        // Some applications (see https://github.com/strimzi/drain-cleaner/issues/34) might send the eviction
+                        // request without the namespace. In such case, we use the namespace form the AdmissionRequest.
+                        LOG.warn("There is no namespace in the Eviction request - trying to use namespace of the Admission request");
+                        namespace = request.getNamespace();
+                    }
```
Comment (scholzj) [inline src/main/java/io/strimzi/ValidatingWebhook.java:98]: Isn't think too late? You already use the namespace to get the pod on line 90.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 20]
```diff
@@ -74,32 +83,35 @@ public AdmissionReview webhook(AdmissionReview review) {
...
+
+                    if (name == null || namespace == null) {
+                        LOG.warn("Failed to decode pod name or namespace from the eviction webhook (pod: {}, namespace: {})", name, namespace);
+                    } else {
+                        LOG.info("Received eviction webhook for Pod {} in namespace {}", name, namespace);
+                        annotatePodForRestart(name, namespace, request.getDryRun());
+                    }
```
Comment (scholzj) [inline src/main/java/io/strimzi/ValidatingWebhook.java:105]: Same as above -> you just used the name / namespace to get the pod.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 21]
```diff
@@ -74,32 +83,35 @@ public AdmissionReview webhook(AdmissionReview review) {
...
                 } else {
-                    LOG.info("Received eviction webhook for Pod {} in namespace {}", name, namespace);
-                    annotatePodForRestart(name, namespace, request.getDryRun());
+                    LOG.info("Received eviction event which does not match any relevant pods.");
                 }
             } else {
-                LOG.info("Received eviction event which does not match any relevant pods.");
+                LOG.warn("no pod has been found with name {} in namespace {}", name, namespace);
```
Comment (scholzj) [inline src/main/java/io/strimzi/ValidatingWebhook.java:110]:
```suggestion
                LOG.warn("No pod has been found with name {} in namespace {}", name, namespace);
```

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 22]
```diff
@@ -302,34 +468,37 @@ public void testNoNamespaceAnywhere() {
         verify(podResource, never()).patch((Pod) any());
     }
 
-    private Pod mockedPod(String podName, boolean kindLabel, boolean ruAnno) {
+
```
Comment (fvaleri) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:471]: Remove 1 extra line here.

File: src/test/java/io/strimzi/ValidatingWebhookTest.java [hunk 23]
```diff
@@ -302,34 +468,37 @@ public void testNoNamespaceAnywhere() {
...
-
         if (ruAnno)  {
             pod.getMetadata().getAnnotations().put("strimzi.io/manual-rolling-update", "true");
         }
 
         return pod;
     }
 
-    private AdmissionReview reviewRequest(String podName, boolean dryRun)   {
+
```
Comment (fvaleri) [inline src/test/java/io/strimzi/ValidatingWebhookTest.java:492]: Remove 2 extra lines here.

**Review by fvaleri — COMMENTED** [review]
Just some nits.

**Review by fvaleri — APPROVED** [review]
LGTM. Thanks, well done!

The pipeline failure is unrelated.

File: src/main/java/io/strimzi/ValidatingWebhook.java [hunk 24]
```diff
@@ -108,41 +118,39 @@ public AdmissionReview webhook(AdmissionReview review) {
...
-                } else {
-                    LOG.info("Pod {} in namespace {} is already annotated for restart", name, namespace);
-                }
-
+    void annotatePodForRestart(Pod pod, boolean dryRun) {
+        String name = pod.getMetadata().getName();
+        String namespace = pod.getMetadata().getNamespace();
+        if (pod.getMetadata() != null
+                && pod.getMetadata().getLabels() != null
+                && "Kafka".equals(pod.getMetadata().getLabels().get("strimzi.io/kind"))) {
```
Comment (scholzj) [inline src/main/java/io/strimzi/ValidatingWebhook.java:126]: Would it make sense to move this check to `matchingLabel` method? You would have all checks in one place. It made sense previously here as we did not get the pod earleir. But now it would make sense to move it. Sorry for not noticing it earlier.

**Review by scholzj — APPROVED** [review]
Thanks for the PR.

**Review by ppatierno — APPROVED** [review]
LGTM
